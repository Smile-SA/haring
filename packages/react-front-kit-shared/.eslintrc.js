module.exports = {
  extends: ['custom'],
  parserOptions: {
    project: ['./tsconfig.json'],
    tsconfigRootDir: __dirname,
  },
  root: true,
};
