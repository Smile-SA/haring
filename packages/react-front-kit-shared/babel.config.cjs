const { babelConfig } = require('test');

module.exports = babelConfig;
