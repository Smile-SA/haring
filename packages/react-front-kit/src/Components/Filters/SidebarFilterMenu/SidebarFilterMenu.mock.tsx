import type { IFiltersItem } from './SidebarFilterMenu';

import { Checkbox, Group } from '@mantine/core';

export const menu: IFiltersItem<number>[] = [
  {
    content: (
      <>
        <Group>
          <Checkbox checked value="Dupont" /> Dupont
        </Group>
        <br />
        <Group>
          <Checkbox checked value="Martin" /> Martin
        </Group>
        <br />
        <Group>
          <Checkbox value="André" /> Andrée
        </Group>
      </>
    ),
    id: 1,
    label: 'Nom du client',
  },
  {
    children: [
      {
        content: (
          <>
            <Group>
              <Checkbox value="CDI" /> CDI
            </Group>
            <br />
            <Group>
              <Checkbox value="CDD" /> CDD
            </Group>
          </>
        ),
        id: 10,
        label: 'Contrat classique',
      },
      {
        content: (
          <Group style={{ padding: '10px 0 10px 10px' }}>
            <Checkbox value="FREELANCE" /> Freelance
          </Group>
        ),
        id: 11,
        label: 'Contrat special',
      },
    ],
    id: 3,
    label: 'Type de contrat',
  },
  {
    id: 8,
    label: 'Période',
  },
  { id: 9, label: 'Durée du contrat' },
];
