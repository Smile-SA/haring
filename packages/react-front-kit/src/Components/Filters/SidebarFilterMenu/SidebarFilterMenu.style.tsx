import { createStyles } from '@mantine/styles';

export const useStyles = createStyles((theme) => ({
  buttonInner: {
    justifyContent: 'Left',
  },
  buttonLabel: {
    color: theme.colors.dark,
    textAlign: 'left',
  },
  buttonRoot: {
    '&[data-selected="false"]': {
      background: 'inherit',
    },
    background: theme.colors.cyan[0],
  },
  contentContainer: {
    padding: '10px',
  },
}));
