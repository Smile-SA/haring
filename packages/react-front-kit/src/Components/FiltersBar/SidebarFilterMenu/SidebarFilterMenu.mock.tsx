import type { IFiltersItem } from './SidebarFilterMenu';

import { Checkbox, Group } from '@mantine/core';

export function getMenu(checked: boolean): IFiltersItem<number>[] {
  return [
    {
      content: (
        <>
          <Group>
            <Checkbox defaultChecked={checked} value="Dupont" /> Dupont
          </Group>
          <br />
          <Group>
            <Checkbox defaultChecked={checked} value="Martin" /> Martin
          </Group>
          <br />
          <Group>
            <Checkbox value="André" /> Andrée
          </Group>
        </>
      ),
      id: 1,
      label: 'Nom du client',
    },
    {
      children: [
        {
          content: (
            <>
              <Group>
                <Checkbox defaultChecked={checked} value="CDI" /> CDI
              </Group>
              <br />
              <Group>
                <Checkbox value="CDD" /> CDD
              </Group>
            </>
          ),
          id: 10,
          label: 'Contrat classique',
        },
        {
          content: (
            <Group style={{ padding: '10px 0 10px 10px' }}>
              <Checkbox defaultChecked={checked} value="FREELANCE" /> Freelance
            </Group>
          ),
          id: 11,
          label: 'Contrat special',
        },
      ],
      id: 3,
      label: 'Type de contrat',
    },
    {
      id: 8,
      label: 'Période',
    },
    { id: 9, label: 'Durée du contrat' },
  ];
}

// Because, the random mantine ids to make fail test
export const menuWithoutContent = [
  {
    id: 1,
    label: 'Nom du client',
  },
  {
    children: [
      {
        id: 10,
        label: 'Contrat classique',
      },
      {
        id: 11,
        label: 'Contrat special',
      },
    ],
    id: 3,
    label: 'Type de contrat',
  },
  {
    id: 8,
    label: 'Période',
  },
  { id: 9, label: 'Durée du contrat' },
];
