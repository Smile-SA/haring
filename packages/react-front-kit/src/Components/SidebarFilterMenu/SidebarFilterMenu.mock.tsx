import type { IFiltersItem } from './SidebarFilterMenu';

import { Checkbox, Group } from '@mantine/core';

export const menu: IFiltersItem<number>[] = [
  {
    content: (
      <>
        <Group>
          <Checkbox value="Dupont" /> Dupont
        </Group>
        <br />
        <Group>
          <Checkbox value="Martin" /> Martin
        </Group>
        <br />
        <Group>
          <Checkbox value="André" /> Andrée
        </Group>
      </>
    ),
    id: 1,
    label: 'Nom du client',
  },
  {
    id: 3,
    label: 'Type de contrat',
  },
  {
    id: 8,
    label: 'Période',
  },
  { id: 9, label: 'Durée du contrat' },
];
