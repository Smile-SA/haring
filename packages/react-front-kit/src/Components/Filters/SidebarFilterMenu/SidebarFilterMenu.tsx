'use client';

import type { ICollapseButtonProps } from '../../CollapseButton/CollapseButton';
import type {
  IMenuItem,
  ISidebarMenuProps,
} from '../../SidebarMenu/SidebarMenu';
import type { PaperProps } from '@mantine/core';
import type { ElementType, ReactElement, ReactNode } from 'react';

import { useMantineTheme } from '@mantine/core';
import { useMemo, useState } from 'react';

import { addPathAndDepth, flattenNestedObjects } from '../../../helpers';
import { CollapseButtonControlled } from '../../CollapseButton/CollapseButtonControlled';

import { useStyles } from './SidebarFilterMenu.style';

export interface IFiltersItem<T extends number | string>
  extends Omit<IMenuItem<T>, 'children'> {
  children?: IFiltersItem<T>[];
  content?: ReactNode;
}
export interface ISidebarFilterMenuProps<
  T extends number | string,
  C extends ElementType,
> extends Omit<ISidebarMenuProps<T, C>, 'menu'>,
    PaperProps {
  menu?: IFiltersItem<T>[];
}

// eslint-disable-next-line react-refresh/only-export-components
export function getRecursiveMenu<
  T extends number | string,
  C extends ElementType,
>(
  classes: {
    buttonInner: string;
    buttonLabel: string;
    buttonRoot: string | undefined;
  },
  setSelectedId: (id?: T) => void,
  onMenuOpen: (id: T, isOpened: boolean) => void,
  openedMenuIds: T[],
  selectedId?: T,
  menu?: IFiltersItem<T>[],
  collapseButtonProps?:
    | Omit<ICollapseButtonProps<T, C>, 'opened'>
    | ((item: IFiltersItem<T>) => Omit<ICollapseButtonProps<T, C>, 'opened'>)
    | undefined,
  level = 0,
): ReactElement[] | null {
  if (!menu || menu.length === 0) {
    return null;
  }
  return menu.map((item) => {
    const { content, children, id, label, leftIcon } = item;
    const theme = useMantineTheme();
    return (
      <CollapseButtonControlled
        key={id}
        classNames={{
          inner: classes.buttonInner,
          label: classes.buttonLabel,
          root: classes.buttonRoot,
        }}
        // prettier-ignore
        collapseStyle={
          openedMenuIds.includes(id) && selectedId === id
            ? {
                backgroundColor: theme.colors.cyan[0],
                borderTop: `1px solid ${theme.colors.cyan[2]}`,
              }
            : openedMenuIds.includes(id)
            ? {
                borderTop: `1px solid ${theme.colors.gray[3]}`,
              }
            : {}
        }
        // prettier-ignore
        id={id}
        isOpenOnSelect
        label={label}
        leftIcon={leftIcon}
        level={level}
        onCollapseChange={(isOpened) => onMenuOpen(id, isOpened)}
        onSelect={setSelectedId}
        opened={openedMenuIds.includes(id)}
        selected={selectedId === id}
        {...(typeof collapseButtonProps === 'function'
          ? collapseButtonProps(item)
          : collapseButtonProps)}
      >
        {getRecursiveMenu(
          classes,
          setSelectedId,
          onMenuOpen,
          openedMenuIds,
          selectedId,
          children,
          collapseButtonProps,
          level + 1,
        )}
        {content}
      </CollapseButtonControlled>
    );
  });
}

export function SidebarFilterMenu<
  T extends number | string,
  C extends ElementType = 'button',
>(props: ISidebarFilterMenuProps<T, C>): ReactElement {
  const {
    collapseButtonProps,
    defaultSelectedId,
    hasOnlyOneOpenMenu = false,
    menu,
    onMenuOpen,
    openedMenuIds,
  } = props;

  const { classes } = useStyles();
  const flatMenu = useMemo(
    () => flattenNestedObjects(addPathAndDepth(menu)),
    [menu],
  );

  const [selectedId, setSelectedId] = useState<T | undefined>(
    defaultSelectedId,
  );

  const [openedIds, setOpenedIds] = useState<T[]>(() => {
    if (defaultSelectedId && !openedMenuIds) {
      const path = flatMenu.find((menu) => menu.id === defaultSelectedId)
        ?.path as T[] | undefined;
      if (path) {
        return path;
      }
    }
    return openedMenuIds ?? [];
  });

  function handleOpenChange(menuId: T, isOpened: boolean): void {
    const openedMenuPath = (flatMenu.find((menu) => menu.id === menuId)?.path ??
      []) as T[];
    onMenuOpen?.(menuId, isOpened, openedMenuPath);
    if (hasOnlyOneOpenMenu && isOpened) {
      setOpenedIds(openedMenuPath);
    } else {
      /** Add or remove id being clicked **/
      const exists = openedIds.includes(menuId);
      let newOpenedIds;
      if (exists) {
        newOpenedIds = openedIds.filter((id) => id !== menuId);
      } else {
        newOpenedIds = openedIds.concat(menuId);
      }
      setOpenedIds(newOpenedIds);
    }
  }

  return (
    <div>
      {getRecursiveMenu(
        classes,
        setSelectedId,
        handleOpenChange,
        openedIds,
        selectedId,
        menu,
        collapseButtonProps,
      )}
    </div>
  );
}
