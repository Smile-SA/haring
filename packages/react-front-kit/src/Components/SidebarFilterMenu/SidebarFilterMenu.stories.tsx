import type { Meta, StoryObj } from '@storybook/react';

import { SidebarFilterMenu as Cmp } from './SidebarFilterMenu';
import { menu } from './SidebarFilterMenu.mock';

const meta = {
  component: Cmp,
  tags: ['autodocs'],
  title: '3-custom/Components/SidebarFilterMenu',
} satisfies Meta<typeof Cmp>;

export default meta;
type IStory = StoryObj<typeof meta>;

export const SidebarFilterMenu: IStory = {
  args: {
    component: 'nav',
    menu,
    openedMenuIds: [],
  },
};
