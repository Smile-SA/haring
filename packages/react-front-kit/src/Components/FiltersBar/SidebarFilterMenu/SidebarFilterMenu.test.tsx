import { renderWithProviders } from '@smile/react-front-kit-shared/test-utils';

import { SidebarFilterMenu } from './SidebarFilterMenu';

describe('SidebarFilterMenu', () => {
  it('matches snapshot', () => {
    const { container } = renderWithProviders(<SidebarFilterMenu />);
    expect(container).toMatchSnapshot();
  });
});
