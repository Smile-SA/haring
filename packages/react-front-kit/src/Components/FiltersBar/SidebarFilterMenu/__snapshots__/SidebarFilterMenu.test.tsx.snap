// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarFilterMenu matches snapshot 1`] = `
<div>
  <div />
</div>
`;
